searx/query.py

```diff
@@ -63,100 +70,225 @@ def _parse_query(self):
+                # this force a engine or category
+                value = query_part[1:].replace('-', ' ').replace('_', ' ')
+                if len(value) > 0:
+                    searx_query_part = self._bang_parse(query_part[0], value, enable_autocomplete)
+                    if searx_query_part and query_part[0] == '!':
+                        self.specific = True
+                elif enable_autocomplete:
+                    self._bang_autocomplete(query_part[0], value)
 
             # append query part to query_part list
-            if searx_query_part:
-                self.query_parts.append(query_part)
+            qlist = self.query_parts if searx_query_part else self.user_query_parts
+
+            qlist.append(query_part)
+            last_index_location = (qlist, len(qlist) - 1)
+
+        self.autocomplete_location = last_index_location
+
+    def _timeout_autocomplete(self):
+        self.autocomplete_list.append('<3')
+        self.autocomplete_list.append('<850')
+
+    def _timeout_parse(self, raw_value):
+        try:
+            raw_timeout_limit = int(raw_value)
+        except ValueError:
+            # error not reported to the user
+            return False
+        else:
+            if raw_timeout_limit < 100:
+                # below 100, the unit is the second ( <3 = 3 seconds timeout )
+                self.timeout_limit = float(raw_timeout_limit)
             else:
-                self.user_query_parts.append(query_part)
+                # 100 or above, the unit is the millisecond ( <850 = 850 milliseconds timeout )
+                self.timeout_limit = raw_timeout_limit / 1000.0
+            return True
+
+    def _language_autocomplete(self, value):
+        if len(value) == 0:
+            # show some example queries
+            self.autocomplete_list.append(":en")
+            self.autocomplete_list.append(":en_us")
+            self.autocomplete_list.append(":english")
+            self.autocomplete_list.append(":united_kingdom")
+        else:
+            for lc in language_codes:
+                lang_id, lang_name, country, english_name = map(str.lower, lc)
+
+                # check if query starts with language-id
+                if lang_id.startswith(value):
+                    if len(value) <= 2:
+                        self.autocomplete_list.append(':' + lang_id.split('-')[0])
+                    else:
+                        self.autocomplete_list.append(':' + lang_id)
+
+                # check if query starts with language name
+                if lang_name.startswith(value) or english_name.startswith(value):
+                    self.autocomplete_list.append(':' + lang_name)
+
+                # check if query starts with country
+                if country.startswith(value.replace('_', ' ')):
+                    self.autocomplete_list.append(':' + country.replace(' ', '_'))
+
+    def _language_parse(self, value, enable_autocomplete):
```

Comment:
This function does too many things. It should not call `_language_autocomlete` because it has nothing to do with configuring autocomplete. Instead please refactor the caller site (from line 83) like this:

```python
if len(value) > 0:
    searx_query_part = self._language_parse(value)
    if not searx_query_part and enable_autocomplete:
        self._language_autocomplete(value)

```

Reply:
Done